content/2016.09.jedit-trackers-with-kinto-en.md

```diff
@@ -0,0 +1,403 @@
+```javascript
+tickets.db.getLastModified().then(function(res) {
+    if (res) {
+        hasDB();
+    } else {
+        noDB();
+    }
+});
+```
+
+To display the server's ticket count before sync, one has to go down to the
+ [kinto-http](https://github.com/Kinto/kinto-http.js) package (this is being worked on, see
+ [this issue](https://github.com/Kinto/kinto-http.js/issues/138)).
+
+code snippet:
+```javascript
+function fetchTotalRecords(api, bucket, name, etag) {
+    var headers = {};
+    headers["Authorization"] = "Basic " + btoa("token:tr0ub4d@ur");
+    return api.client.execute({
+        path: "/buckets/" + bucket + "/collections/" + name + "/records" + (etag ? ("?_since=" + etag) : ""),
+        method: "HEAD",
+        headers
+    }, {
+        raw: true
+    }).then(function(res) {
+        console.log("RES", res);
+        return parseInt(res.headers.get("total-records"));
+    });
+}
+
+tickets.db.getLastModified().then(function(lastMod) {
+    var api = tickets.api.bucket(tickets.bucket).collection(tickets.name);
+    // you'll soon be able to replace it with api.getLastModified(lastMod)
+    return fetchTotalRecords(api, tickets.bucket, tickets.name, lastMod);
+}).then(renderTicketCount);
+```
+
+This method also works to find out if updated tickets exist on the server
+( [via the ETag](http://kinto.readthedocs.io/en/latest/api/1.x/filtering.html#polling-for-changes)).
+The kinto server doesn't respond instantly as it should. Maybe I should also put the ETag in an `If-None-Match`
+header to speed things up.
+
+## Searching for Tickets
+
+I've used [lunr](http://lunrjs.com/) to implement client-side full text search. It builds a Lucene-like index and
+then lets one query this index. The index is serialized and saved as a document in a local Kinto.js collection
+to be reused on next page load.
+
+Searching the index returns an array of document ids, from which I load documents before filtering the resulting
+(usually small) array with remaining query params. This is performing better than the first version of the
+application, when I first loaded all documents by calling `tickets.list()`.
+
+To search for *hello* in the lunr index:
+```javascript
+> index.search("hello")
+   [{"ref":"52a4d8cb27184667e3400f42","score":0.019459983366287688},
+    {"ref":"5690ff3f24b0d96f08df7d6a","score":0.013994677485991343},
+    ...
+   ]
+```
+
+To get a `{ data: ['array', 'of', 'tickets', 'containing', 'myText'] }`  Promise:
+```javascript
+return Promise.all(index.search(myText).map(function(hit) {
+    return tickets.get(hit.ref);
+})).then(function(allRes) {
+    return {
+        data: allRes.map(function(res) {
+            return res.data;
+        })
+    };
+});
+```
+
+Matching documents are fetched from IndexedDB in parallel via their Id (`collection.get()`).
+It's fast because IndexedDB is optimised for key based access, but it could be more efficient to use
+`tickets.list({id: ['array','of ids']})` to curb request count.
```

Comment:
Obviously this is just a silly example, but should this be `['array','of','ids']` ? :)